src/globus_sdk/experimental/globus_app/globus_app.py

```diff
@@ -0,0 +1,429 @@
+        authorize requests.
+        """
+
+    @abc.abstractmethod
+    def run_login_flow(
+        self, auth_params: GlobusAuthorizationParameters | None = None
+    ) -> None:
+        """
+        Run an authorization flow to get new tokens which are stored and available
+        for the next authorizer gotten by get_authorizer.
+
+        :param auth_params: A GlobusAuthorizationParameters to control how the user
+            will authenticate. If not passed
+        """
+
+    def _auth_params_with_required_scopes(
+        self,
+        auth_params: GlobusAuthorizationParameters | None = None,
+    ) -> GlobusAuthorizationParameters:
+        """
+        Either make a new GlobusAuthorizationParameters with this app's required scopes
+        or combine this app's required scopes with given auth_params.
+        """
+        required_scopes = []
+        for scope_list in self._scope_requirements.values():
+            required_scopes.extend(scope_list)
+
+        if auth_params:
+            if auth_params.required_scopes:
+                combined_scopes = Scope.merge_scopes(
+                    required_scopes, [Scope(s) for s in auth_params.required_scopes]
+                )
+                auth_params.required_scopes = [str(s) for s in combined_scopes]
+            else:
+                auth_params.required_scopes = [str(s) for s in required_scopes]
+            return auth_params
+
+        else:
+            return GlobusAuthorizationParameters(
+                required_scopes=[str(s) for s in required_scopes]
+            )
+
+    def get_authorizer(self, resource_server: str) -> GlobusAuthorizer:
+        """
+        Get a ``GlobusAuthorizer`` from the app's authorizer factory for a specified
+        resource server. The type of authorizer is dependent on the app.
+
+        :param resource_server: the resource server the Authorizer will provide
+            authorization headers for
+        """
+        return self._authorizer_factory.get_authorizer(resource_server)
+
+    def add_scope_requirements(
+        self, scope_requirements: dict[str, list[Scope]]
+    ) -> None:
+        """
+        Add given scope requirements to the app's scope requirements.
+
+        :param scope_requirements: a dict of Scopes indexed by resource server
+            that will be added to this app's scope requirements
+        """
+        for resource_server, scopes in scope_requirements.items():
+            if resource_server not in self._scope_requirements:
+                self._scope_requirements[resource_server] = scopes
+
+            else:
+                self._scope_requirements[resource_server] = Scope.merge_scopes(
+                    self._scope_requirements[resource_server], scopes
+                )
+
+        # update the app's ValidatingTokenStorage's scope requirements to enforce the
+        # updated scope requirements
+        self._validating_token_storage.scope_requirements = self._scope_requirements
```

Comment:
This shouldn't be necessary. Scope requirements are passed by reference; so an update to them here will mean an update to them there.